mitodl/mitxpro
# feat: view courses by topic

#### Pre-Flight checklist


#### What are the relevant tickets?
Epic: https://github.com/mitodl/mitxpro/issues/2589
Subtask (HomePage Changes): https://github.com/mitodl/mitxpro/issues/2590
Subtask (CatalogPage Changes): https://github.com/mitodl/mitxpro/issues/2591
Subtask (Adds topics to CMS and associates with CoursePage): https://github.com/mitodl/mitxpro/issues/2593

#### What's this PR do?
This is the main PR for https://github.com/mitodl/mitxpro/issues/2589.
All of the subtask PRs are tested and reviewed. Here is the list of subtask PRs:
https://github.com/mitodl/mitxpro/pull/2594
https://github.com/mitodl/mitxpro/pull/2596
https://github.com/mitodl/mitxpro/pull/2597

Here are the changes this PR adds:

- Makes topics model to support multi-level hierarchy
- Enables topics management through CMS
- Allow CMS pages to allow associating topics to a course
- Migrates existing course topics to course pages
- Fixes the issue in the seed data command
- Adds course topic filters on the xPro HomePage
- Adds a course topic filter on the catalog page
- Updates banner and featured product position

#### How should this be manually tested?

PART 1 (CMS, Django Admin):
- On the master branch, set up a few courses. The `seed_data` should be fine
- Shift to this branch, and check all the migrations ran OK
- Check that your CMS course pages now have an associated topic from the Django Course model
- Open `/cms` and click `Course Topics`, Make sure you see the topic list, can edit and create new topics from here
- Open a course page in CMS, and change its topic. Make sure it gets changed successfully and also syncs with the Course model in Django - This means that the course model in Django and CoursePage in CMS should have the same topics associated all the time. (Maybe we should completely remove topics from the Course model in Django)
- Try to run `./manage.py seed_data` and see that it works fine
- You can also `docker-compose down` and then `docker-compose up` to run `./manage.py seed_data` to see how it works on a fresh instance - This should also work fine.

PART 2 (Catalog Page):
- Verify the designs
- Verify that the filters are working fine
  - Visit Wagtail and create some parent course topics having parent=null and some child topics with parents.
  - Visit Wagtail admin, Visit course pages, and Add some topics.
  - Parent topics should work fine, If a child topic is added to a course it should appear when the parent filter is applied.
  - Same should be verified for programs, A program having courses with the filtered title should appear in the catalog.

PART 3 (HomePage):
- Visit the xPro Homepage.
- Make sure that the changes are as per the designs.
- Make sure that the links redirect to the catalog page.
- Slider is visible for all screens when topics are more than 6.
- Slider is visible for small screens when topics are 6 or fewer.


#### Screenshots (if appropriate)
**Topics list view in Wagtail**
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/34372316/228163688-57d182f3-6da2-4109-99d4-2cf28d7d14c4.png">

**Topics Edit View**

<img width="1164" alt="image" src="https://user-images.githubusercontent.com/34372316/228163864-00001885-4e98-49ec-820d-f57aad54f193.png">

**Topics Add view in Wagtail**
<img width="969" alt="image" src="https://user-images.githubusercontent.com/34372316/228164017-a29c1b6f-e38e-4732-a986-cae08cd23afa.png">

**Topics selection view in course**
<img width="1590" alt="image" src="https://user-images.githubusercontent.com/34372316/228164843-39da3239-dc16-4e97-a1ba-7c982ba3456f.png">

**Delete view for topic**
<img width="740" alt="image" src="https://user-images.githubusercontent.com/34372316/228458617-f4beb68a-9b09-4ab8-bc5e-7b905cf1a699.png">


HOMEPAGE:
<img width="1441" alt="Screen Shot 2023-03-29 at 1 19 41 PM" src="https://user-images.githubusercontent.com/52656433/228472355-7dd64e8c-7330-4d04-84ca-05d8a4d2b967.png">
<img width="1024" alt="Screen Shot 2023-03-29 at 1 19 56 PM" src="https://user-images.githubusercontent.com/52656433/228472374-c99b02aa-1cf0-43a2-9721-79d89b8ba239.png">
<img width="772" alt="Screen Shot 2023-03-29 at 1 20 08 PM" src="https://user-images.githubusercontent.com/52656433/228472385-e304e610-6763-4b64-9b15-dfb05d067d23.png">
<img width="429" alt="Screen Shot 2023-03-29 at 1 20 24 PM" src="https://user-images.githubusercontent.com/52656433/228472409-39effc77-cb67-438a-822f-cd9dd36ab9ec.png">
<img width="1439" alt="Screen Shot 2023-03-29 at 1 21 03 PM" src="https://user-images.githubusercontent.com/52656433/228472418-26f9b3e5-e4db-4807-870d-90723870b420.png">
<img width="1023" alt="Screen Shot 2023-03-29 at 1 21 20 PM" src="https://user-images.githubusercontent.com/52656433/228472424-6d7a4661-fc0a-4546-b8ec-492a24258d2d.png">
<img width="767" alt="Screen Shot 2023-03-29 at 1 21 33 PM" src="https://user-images.githubusercontent.com/52656433/228472438-0d86c7e7-bfd9-4c27-abf1-093ac1578ba9.png">
<img width="424" alt="Screen Shot 2023-03-29 at 1 21 45 PM" src="https://user-images.githubusercontent.com/52656433/228472449-46604747-c5f7-4d69-aec3-d4aba9c046af.png">


CATALOG PAGE:

Laptop 1440:
<img width="1753" alt="Screen Shot 2023-03-22 at 5 55 29 PM" src="https://user-images.githubusercontent.com/52656433/226911259-436c3e19-3ba3-4d85-b788-4775e87fed21.png">

Laptop 1024:
<img width="1026" alt="Screen Shot 2023-03-22 at 5 55 57 PM" src="https://user-images.githubusercontent.com/52656433/226911378-fe4faad6-a0ad-47fa-a334-e4450b23c97b.png">

Tablet:
<img width="770" alt="Screen Shot 2023-04-06 at 3 26 59 PM" src="https://user-images.githubusercontent.com/52656433/230350999-e374ee28-bfe1-426a-843f-fe885f14e001.png">

Mobile L:
<img width="429" alt="Screen Shot 2023-04-06 at 3 27 10 PM" src="https://user-images.githubusercontent.com/52656433/230350934-5b4dc712-0670-48d1-b6ed-7f3165c900d1.png">
